weighing.py

```diff
@@ -392,10 +432,126 @@ def create_quality_test(self):
+        Invoice = pool.get('account.invoice')
+        Company = pool.get('company.company')
+        context = Transaction().context
+
+        invoices = Invoice.search([
+            ('party', '=', party),
+            ('state', '=', 'draft'),
+            ('type', '=', 'in')])
+        if invoices:
+            invoice = invoices[0]
+        if not invoices:
+            journals = Journal.search([
+                ('type', '=', 'expense'),
+                ], limit=1)
+            if journals:
+                journal, = journals
+            else:
+                journal = None
+
+            invoice = Invoice()
+            invoice.company = Company(context['company'])
+            invoice.type = 'in'
+            invoice.journal = journal
+            invoice.party = party
+            invoice.invoice_address = party.address_get(type='invoice')
+            invoice.currency = Company(context['company']).currency
+            invoice.account = party.account_payable_used
+            invoice.payment_term = party.supplier_payment_term
+        return invoice
+
     @classmethod
     @Workflow.transition('done')
     def done(cls, weighings):
-        pass
+        pool = Pool()
+        InvoiceLine = pool.get('account.invoice.line')
+        Product = pool.get('product.product')
+        Company = pool.get('company.company')
+        context = Transaction().context
+        ContractProductPriceListTypePriceList = pool.get(
+            'agronomics.contract-product.price_list.type-product.price_list')
+        WeighingInvoiceLine = pool.get(
+            'agronomics.weighing-account.invoice.line')
+        RecomputeCostPrice = pool.get('product.recompute_cost_price',
+            type='wizard')
+        Move = pool.get('stock.move')
+
+        default_invoice_line_values = InvoiceLine.default_get(
+            InvoiceLine._fields.keys(), with_rec_name=False)
+        invoice_line = InvoiceLine(**default_invoice_line_values)
+
+        to_save = []
+        to_save_moves = []
+        to_recompute_products = []
+        for weighing in weighings:
+            cost_price = Decimal(0)
+            if weighing.beneficiaries:
+                for beneficiary in weighing.beneficiaries:
+                    price_list = ContractProductPriceListTypePriceList.search([
+                        ('contract', '=', weighing.purchase_contract),
+                        ('price_list_type', '=',
+                            beneficiary.product_price_list_type),
+                    ])
+
+                    invoice_line = InvoiceLine()
+                    invoice_line.type = 'line'
+                    invoice_line.invoice_type = 'in'
+                    invoice_line.party = beneficiary.party
+                    invoice_line.currency = (
+                        Company(context['company']).currency)
+                    invoice_line.company = Company(context['company'])
+                    invoice_line.description = ''
```

Comment:
Same here.
invoice_line.product = product
invoice_line.on_change_product()  
(and you avoid to fill unit, accounts ... )
invoice_line.quantity = weighing.netweight 
invoice_line.on_change_quantity()
(and you avoid to calc unit_price .. )

Reply:
Ok for the "on_change_product()" change. But we can't use the "on_change_quantity" to calculate the unit_price.
The unit price is different for each beneficiary. We calculate the unit_price using the product_price_list_type.